components/cloud/src/blob.rs

```diff
@@ -0,0 +1,192 @@
+use futures_io::AsyncRead;
+pub use kvproto::backup::CloudDynamic;
+use std::io;
+use std::marker::Unpin;
+
+/// An abstraction for blob storage.
+/// Currently the same as ExternalStorage
+pub trait BlobStorage: 'static + Send + Sync {
+    fn name(&self) -> &'static str;
+
+    fn url(&self) -> io::Result<url::Url>;
+
+    /// Write all contents of the read to the given path.
+    fn write(
+        &self,
+        name: &str,
+        reader: Box<dyn AsyncRead + Send + Unpin>,
+        content_length: u64,
+    ) -> io::Result<()>;
```

Comment:
Maybe we can make the io methods under BlobStorage async. It seems that async method are required for implementing the grpcio service and its better to usage **BlobStorage** trait directly. When tikv want use grcp to read/write file, it should use an object implemented the **ExternalStorage** by wrapping a grpcio client to make it.

Reply:
This is implemented now in the latest commit. I did not convert the methods to be async because the compiler didn't like that idea, but I hope we will eventually.